core/src/worker/heartbeat.rs

```diff
@@ -0,0 +1,224 @@
+            }
+            Some(PbDuration {
+                seconds: dur.as_secs() as i64,
+                nanos: dur.subsec_nanos() as i32,
+            })
+        } else {
+            None
+        };
+
+        self.heartbeat_time = Some(now);
+
+        self.reset_notify.notify_one();
+
+        Some(WorkerHeartbeat {
+            worker_instance_key: self.worker_instance_key.clone(),
+            worker_identity: self.worker_identity.clone(),
+            host_info: Some(self.host_info.clone()),
+            task_queue: self.task_queue.clone(),
+            sdk_name: self.sdk_name.clone(),
+            sdk_version: self.sdk_version.clone(),
+            status: 0,
+            start_time: Some(self.start_time.into()),
+            heartbeat_time: Some(SystemTime::now().into()),
+            elapsed_since_last_heartbeat,
+            ..Default::default()
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_help::WorkerExt;
+    use crate::test_help::test_worker_cfg;
+    use crate::worker;
+    use crate::worker::client::mocks::mock_worker_client;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use temporal_sdk_core_api::worker::PollerBehavior;
+    use temporal_sdk_core_protos::temporal::api::workflowservice::v1::RecordWorkerHeartbeatResponse;
+
+    #[tokio::test]
+    async fn worker_heartbeat() {
+        let mut mock = mock_worker_client();
+        mock.expect_record_worker_heartbeat()
+            .times(2)
+            .returning(move |heartbeat| {
+                let host_info = heartbeat.host_info.clone().unwrap();
+                assert_eq!("test-identity", heartbeat.worker_identity);
+                assert!(!heartbeat.worker_instance_key.is_empty());
+                assert_eq!(
+                    host_info.host_name,
+                    gethostname::gethostname().to_string_lossy().to_string()
+                );
+                assert_eq!(host_info.process_id, std::process::id().to_string());
+                assert_eq!(heartbeat.sdk_name, "test-core");
+                assert_eq!(heartbeat.sdk_version, "0.0.0");
+                assert_eq!(heartbeat.task_queue, "q");
+                assert!(heartbeat.heartbeat_time.is_some());
+                assert!(heartbeat.start_time.is_some());
+
+                Ok(RecordWorkerHeartbeatResponse {})
+            });
+
+        let config = test_worker_cfg()
+            .activity_task_poller_behavior(PollerBehavior::SimpleMaximum(1_usize))
+            .max_outstanding_activities(1_usize)
+            .heartbeat_interval(Duration::from_millis(200))
+            .build()
+            .unwrap();
+
+        let heartbeat_fn = Arc::new(OnceLock::new());
+        let client = Arc::new(mock);
+        let worker = worker::Worker::new(config, None, client, None, Some(heartbeat_fn.clone()));
+        heartbeat_fn.get().unwrap()();
+
+        // heartbeat timer fires once
+        tokio::time::sleep(Duration::from_millis(300)).await;
```

Comment:
Looks like we can avoid the sleeps by using `advance` and family which are behind `test-utils` flags for tokio: https://docs.rs/tokio/latest/tokio/time/index.html

Which is very nice.